x-pack/metricbeat/module/prometheus/remote_write/data.go

```diff
@@ -0,0 +1,272 @@
+				if !ok {
+					hist = histogram{}
+				}
+				hist.buckets = append(hist.buckets, b)
+				hist.timestamp = metric.Timestamp.Time()
+				hist.labels = labelsClone
+				hist.metricName = name
+				histograms[histKey] = hist
+				continue
+			}
+			e.ModuleFields.Update(data)
+		}
+	}
+
+	// process histograms together
+	g.processPromHistograms(eventList, histograms)
+	return eventList
+}
+
+// rateCounterUint64 fills a counter value and optionally adds the rate if rate_counters is enabled
+func (g *remoteWriteTypedGenerator) rateCounterUint64(name string, labels common.MapStr, value uint64) common.MapStr {
+	d := common.MapStr{
+		"counter": value,
+	}
+
+	if g.rateCounters {
+		d["rate"], _ = g.counterCache.RateUint64(name+labels.String(), value)
+	}
+
+	return d
+}
+
+// rateCounterFloat64 fills a counter value and optionally adds the rate if rate_counters is enabled
+func (g *remoteWriteTypedGenerator) rateCounterFloat64(name string, labels common.MapStr, value float64) common.MapStr {
+	d := common.MapStr{
+		"counter": value,
+	}
+	if g.rateCounters {
+		d["rate"], _ = g.counterCache.RateFloat64(name+labels.String(), value)
+	}
+
+	return d
+}
+
+// processPromHistograms receives a group of Histograms and converts each one to ES histogram
+func (g *remoteWriteTypedGenerator) processPromHistograms(eventList map[string]mb.Event, histograms map[string]histogram) {
+	for _, histogram := range histograms {
+		labelsHash := histogram.labels.String()
+		if _, ok := eventList[labelsHash]; !ok {
+			eventList[labelsHash] = mb.Event{
+				ModuleFields: common.MapStr{},
+			}
+
+			// Add labels
+			if len(histogram.labels) > 0 {
+				eventList[labelsHash].ModuleFields["labels"] = histogram.labels
+			}
+		}
+
+		e := eventList[labelsHash]
+		e.Timestamp = histogram.timestamp
+
+		hist := dto.Histogram{
+			Bucket: histogram.buckets,
+		}
+		name := strings.TrimSuffix(histogram.metricName, "_bucket")
+		data := common.MapStr{
+			name: common.MapStr{
+				"histogram": collector.PromHistogramToES(g.counterCache, histogram.metricName, histogram.labels, &hist),
+			},
+		}
+		e.ModuleFields.Update(data)
+	}
+}
+
+// findMetricType evaluates the type of the metric by check the metricname format in order to handle it properly
+func (g *remoteWriteTypedGenerator) findMetricType(metricName string, labels common.MapStr) string {
+	leLabel := false
+	if _, ok := labels["le"]; ok {
+		leLabel = true
+	}
+	if strings.HasSuffix(metricName, "_total") || strings.HasSuffix(metricName, "_sum") ||
+		strings.HasSuffix(metricName, "_count") {
+		return counterType
+	} else if strings.HasSuffix(metricName, "_bucket") && leLabel {
+		return histogramType
+	}
+
+	// handle user provided patterns
+	if len(g.counterPatterns) > 0 {
+		if p.MatchMetricFamily(metricName, g.counterPatterns) {
+			return counterType
+		}
+	}
+	if len(g.histogramPatterns) > 0 {
+		if p.MatchMetricFamily(metricName, g.histogramPatterns) && leLabel {
+			return histogramType
+		}
+	}
```

Comment:
Would it make sense to check user-provided patterns before the fixed ones? In case someone wants to override default behaviour for an specific metric.

Reply:
I think this is a valid point 👍 . I will make it like this.